api/src/main/java/org/pmiops/workbench/workspaces/WorkspaceServiceImpl.java

```diff
@@ -411,24 +412,26 @@ public void updateWorkspaceBillingAccount(DbWorkspace workspace, String newBilli
       fireCloudService.updateBillingAccountAsService(
           workspace.getWorkspaceNamespace(), newBillingAccountName);
     } else {
-      try {
-        Optional<Boolean> isOpenMaybe =
-            Optional.ofNullable(
-                cloudBillingClient.getBillingAccount(newBillingAccountName).getOpen());
-        boolean isOpen = isOpenMaybe.orElse(false);
-        if (!isOpen) {
-          throw new BadRequestException(
-              "Provided billing account is closed. Please provide an open account.");
-        }
-        fireCloudService.updateBillingAccount(
-            workspace.getWorkspaceNamespace(), newBillingAccountName);
-      } catch (IOException e) {
-        throw new ServerErrorException("Could not fetch user provided billing account.", e);
+      fireCloudService.updateBillingAccount(
+          workspace.getWorkspaceNamespace(), newBillingAccountName);
+    }
+
+    try {
+      ProjectBillingInfo projectBillingInfo =
+          cloudBillingClient.pollUntilBillingAccountLinked(
+              newBillingAccountName, workspace.getGoogleProject());
+      if (!projectBillingInfo.getBillingEnabled()) {
+        throw new BadRequestException(
```

Comment:
Does Terra allow us to link that? I guess it's odd that the linkage actually succeeded here, but we throw a 400 anyways.

Existing issue but I prefer 412 for these cases - when a request is well formed, but the current system state does not permit the operation (and of course - the other 400 codes are not applicable). For example, if I fixed the billing account and tried the request again, it would succeed - to me this generally means 400 is not appropriate.